### PR TITLE
Remove need to store MPI communicator in PETScWrappers::SolverBase

### PR DESCRIPTION
**Pull request by bangerth**
This addresses the first part of the concerns raised in #14428 (@marcfehling @kronbichler FYI). The patch removes the need to store the MPI communicator in the PETSc `SolverBase` class and instead gets the information from other places. 

There is a small hiccup in the scheme in that, for some use cases with SLEPc, we need to be able to query the MPI communicator from the PETSc `PreconditionBase` class. That class does not currently store this information, but fortunately, all derived classes have this information and can provide it. So this does not present a problem from an interface perspective.

This patch only addresses the issue of whether PETSc solvers store a the communicator. They now no longer do, and consequently, the separate communicators given to their constructors are ignored. I will change the constructor calls in a second step to remove the communicator object, but thought I'd post this patch here already.

The patch passes all PETSc and SLEPc tests on my system.

/rebuild

### CODE REVIEW

**Review by marcfehling — APPROVED** [review]
Looks good! Part of #14426.

**Comment by bangerth** [comment]
I've got the rest of the patch ready but have to run now. Feel free to merge this part already if it tests alright, and I'll push the rest tomorrow/later today!

**Comment by marcfehling** [comment]
I didn't know which version of petsc/slepc you tested with. I double-checked with `3.17.1` that I have installed, and the (non-complex) petsc and slepc tests pass with this patch.

**Comment by bangerth** [comment]
I'm using PETSc 3.7.4 and it also worked without problem. So I'm pretty sure this works as expected.

**Comment by bangerth** [comment]
I ended up pushing the missing commits that convert all places I could find to not pass a communicator in the constructor. This includes in step-40, as well as a changelog entry.

File: source/lac/petsc_precondition.cc [hunk 1]
```diff
@@ -960,13 +1024,19 @@ namespace PETScWrappers
...
   void
   PreconditionBDDC<dim>::initialize(const MatrixBase &    matrix_,
                                     const AdditionalData &additional_data_)
   {
     clear();
 
+    mpi_communicator = matrix_.get_mpi_communicator();
+
+    mpi_communicator = matrix_.get_mpi_communicator();
+
```
Comment (marcfehling) [inline source/lac/petsc_precondition.cc:1039]:
```suggestion
```

duplicate

**Review by marcfehling — APPROVED** [review]
Looks good. Found one copy&paste issue.

**Review by kronbichler — APPROVED** [review]
Nice!

**Comment by bangerth** [comment]
I fixed the duplicate statement.

File: include/deal.II/lac/petsc_precondition.h [hunk 2]
```diff
@@ -62,7 +62,7 @@ namespace PETScWrappers
     /**
      * Constructor.
      */
-    PreconditionBase();
+    PreconditionBase(const MPI_Comm &mpi_communicator = MPI_COMM_NULL);
```
Comment (masterleinad) [inline include/deal.II/lac/petsc_precondition.h:65]: I think we should deprecate the old constructor then.

Comment (tjhei) [reply include/deal.II/lac/petsc_precondition.h:65]: Yes, and it should probably be made explicit?

Comment (bangerth) [reply include/deal.II/lac/petsc_precondition.h:65]: I don't think this is really necessary. This is, for all intents and purposes an internal class. It is of course conceivable that some user is wrapping another PETSc preconditioner, but I think it is not particularly likely. If you insist, I can put the old constructor back in, of course.

**Comment by bangerth** [comment]
It was easier to just put that constructor back in than to discuss the issue at length :-) I pushed the corresponding patch.

**Review by masterleinad — APPROVED** [review]

**Pull request merged**